### PR TITLE
feat: scope search filters to the browsed category and collection

### PR DESCRIPTION
**Pull request by gajda-w**
The facets query asked for every storefront attribute and discarded the
search parameters it was given, so the filters menu was identical on every
surface. It now narrows the attribute query to the browsed categories and
collections, and re-reads the facets when a filter settles so the menu
narrows before the shopper submits it.

The category and collection filters are built from the catalog instead of
from hand-maintained Saleor product attributes mirroring it, so a category
or collection becomes filterable as soon as it exists. Both keys are
reserved across providers, so an attribute sharing one of those slugs
cannot shadow the real filter.

# Pull Request Checklist

- [ ] Target `main` from a short-lived change branch and use a Conventional Commit PR title.
- [ ] Test the changes locally to ensure they work as expected.
- [ ] Document the testing process and results in the pull request description. (Screen recording, screenshot etc)
- [ ] Verify all four Vercel project statuses: `nimara-docs`, `nimara-ecommerce`,
      `nimara-ecommerce-stripe`, and `nimara-marketplace`.
- [ ] Include new tests for any new functionality or significant changes.
- [ ] Ensure that tests cover edge cases and potential failure points.
- [ ] Keep incomplete behavior disabled with a short-lived feature flag or branch-by-abstraction seam.
- [ ] Document the impact of the changes on the system, including potential risks and benefits.
- [ ] Provide a rollback plan in case the changes introduce critical issues.
- [ ] Update documentation to reflect any changes in functionality.

### CODE REVIEW

File: packages/features/src/search/shared/components/filters/filters-container.tsx [hunk 1]
```diff
@@ -37,62 +49,121 @@ type Props = {
...
+   * Every server action re-renders the route, so a new `facets` prop is not by
+   * itself fresher than what this panel already fetched. Only the applied
+   * filters changing makes the local set stale.
+   */
+  if (appliedKey !== syncedKey) {
+    setSyncedKey(appliedKey);
+    setFacets(serverFacets);
+    fetchedForRef.current = appliedKey;
+    dropdownValuesRef.current = {};
+  }
```
Comment (piotrgrundas) [inline packages/features/src/search/shared/components/filters/filters-container.tsx]: I think this should be wrapped in useEffect

File: packages/features/src/search/shared/components/filters/filters-container.tsx [hunk 2]
```diff
@@ -37,62 +49,121 @@ type Props = {
...
+            onCommit={refreshFacets}
+            onValueChange={handleDropdownValueChange}
+          />
+        );
+      case "PLAIN_TEXT":
+        return <FilterText key={facet.slug} facet={facet} value={value} />;
+      case "SWATCH":
+        return <ColorSwatch key={facet.slug} facet={facet} value={value} />;
+    }
+  };
```
Comment (piotrgrundas) [inline packages/features/src/search/shared/components/filters/filters-container.tsx:141]: IMO it would be better to extract those functions outside render body.

File: apps/storefront/src/app/[locale]/(main)/_actions/get-facets.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  categorySlug: string | undefined,
+  { query, filters }: { filters: Record<string, string>; query: string },
+): Promise<Facet[]> {
+  const [services, region] = await Promise.all([
+    getServiceRegistry(),
+    getCurrentRegion(),
+  ]);
+
+  return fetchFacets({ services, region, query, filters, categorySlug });
+}
```
Comment (piotrgrundas) [inline apps/storefront/src/app/[locale]/(main)/_actions/get-facets.ts]: Wonder if it shouldn't be in sth like `packages/features/src/search/shared/actions.ts`?

File: packages/features/src/search/shared/helpers/filters.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * Params that address the result set instead of narrowing it, so neither the
+ * product search nor the facet lookup receives them as filters.
+ */
+const NON_FILTER_PARAMS: string[] = [
+  ...PASS_THROUGH_PARAMS,
+  ...PAGINATION_PARAMS,
+];
+
+export const getFiltersFromSearchParams = (
```
Comment (piotrgrundas) [inline packages/features/src/search/shared/helpers/filters.ts:20]: Missing tests for this.

File: packages/infrastructure/src/search/algolia/helpers.ts [hunk 5]
```diff
@@ -37,6 +41,12 @@ export const getIndexName = (
   return replica.indexName;
 };
 
+const toCategoryName = (slug: string) => {
+  const name = slug.replaceAll("-", " ");
+
+  return name.charAt(0).toUpperCase() + name.slice(1);
+};
```
Comment (piotrgrundas) [inline packages/infrastructure/src/search/algolia/helpers.ts]: This might bite. Had issues in the past using transformations like that, and it might lead into wrong filtering.
I'm assuming you want to create an Algolia filter by category, which is a name. Ideally, it should be passed as it's received from Algolia.

File: packages/infrastructure/src/search/saleor/infrastructure/get-facets-infra.ts [hunk 6]
```diff
@@ -1,48 +1,129 @@
...
+      apiURI: apiURL,
+      logger,
+    });
+    const cacheOptions: FetchOptions = {
+      next: {
+        // FIXME: Temp value for now
+        revalidate: 5 * 60,
+        tags: ["SEARCH", "SEARCH:FACETS"],
+      },
+    };
```
Comment (piotrgrundas) [inline packages/infrastructure/src/search/saleor/infrastructure/get-facets-infra.ts]: Shouldn't options be passed as an object to infra to anyone using it? 
Infra should be framework-agnostic.

File: packages/infrastructure/src/search/saleor/infrastructure/get-facets-infra.ts [hunk 7]
```diff
@@ -1,48 +1,129 @@
...
+         * cache entry.
+         */
+        client.execute(FacetsTaxonomyQueryDocument, {
+          variables: {
+            channel: context.channel,
+            languageCode: context.languageCode as LanguageCodeEnum,
+            categoryLevel: ROOT_CATEGORY_LEVEL,
+          },
+          options: cacheOptions,
+          operationName: "FacetsTaxonomyQuery",
```
Comment (piotrgrundas) [inline packages/infrastructure/src/search/saleor/infrastructure/get-facets-infra.ts]: I think I made `operationName` deprecated and added auto-discovery of the used operation.

File: packages/infrastructure/src/search/saleor/infrastructure/get-facets-infra.ts [hunk 8]
```diff
@@ -1,48 +1,129 @@
...
+        /**
+         * A vendor-owned collection groups one vendor's own products, and the
+         * marketplace mints one per vendor, so listing them store-wide would
+         * grow without bound.
+         */
+        .filter(({ node }) => !node.vendorId?.trim())
+        .map(({ node }) => ({
+          name: getTranslation("name", node),
+          slug: node.slug,
+        })) ?? [];
```
Comment (piotrgrundas) [inline packages/infrastructure/src/search/saleor/infrastructure/get-facets-infra.ts:79]: Maybe a common Sleor helper for excluding vendor collections? I imagine it might be used in other places.

File: packages/infrastructure/src/search/saleor/helpers.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+          })),
+        },
+      ]
+    : [];
+
+export const buildCategoryFacet = (categories: TaxonomyEntry[]): Facet[] =>
+  buildTaxonomyFacet(CATEGORY_FILTER_KEY, "filters.category", categories);
+
+export const buildCollectionFacet = (collections: TaxonomyEntry[]): Facet[] =>
+  buildTaxonomyFacet(COLLECTION_FILTER_KEY, "filters.collections", collections);
```
Comment (piotrgrundas) [inline packages/infrastructure/src/search/saleor/helpers.ts:68]: Test this file please.

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

**Review by piotrgrundas — APPROVED** [review]
